potoken.py

```diff
@@ -0,0 +1,255 @@
+            return False
+        if self._update_requested.is_set():
+            logger.debug('force update has already been requested')
+            return False
+        self._loop.call_soon_threadsafe(self._update_requested.set)
+        logger.debug('force update requested')
+        return True
+
+    @staticmethod
+    def _extract_token(request: nodriver.cdp.network.Request) -> Optional[dict]:
+        post_data = request.post_data
+        try:
+            post_data_json = json.loads(post_data)
+            visitor_data = post_data_json['context']['client']['visitorData']
+            potoken = post_data_json['serviceIntegrityDimensions']['poToken']
+        except (json.JSONDecodeError, TypeError, KeyError) as e:
+            logger.warning(f'failed to extract token from request: {type(e)}, {e}')
+            return None
+        token_info = {
+            'updated': int(time.time()),
+            'potoken': potoken,
+            'visitor_data': visitor_data
+        }
+        return token_info
+
+    async def _update(self) -> None:
+        try:
+            await asyncio.wait_for(self._perform_update(), timeout=600)
+        except asyncio.TimeoutError:
+            logger.error(f'update failed: hard limit timeout exceeded. Browser might be failing to start properly')
+
+    async def _perform_update(self) -> None:
+        if self._ongoing_update.locked():
+            logger.debug('update is already in progress')
+            return
+
+        async with self._ongoing_update:
+            logger.info(f'update started')
+            self._extraction_done.clear()
+
+            browser = await nodriver.start(headless=False, user_data_dir=self.profile_path)
+            tab = browser.main_tab
+            tab.add_handler(nodriver.cdp.network.RequestWillBeSent, self._send_handler)
+            await tab.get('https://www.youtube.com/embed/jNQXAC9IVRw')
+            player_clicked = await self._click_on_player(tab)
+            if player_clicked:
+                await self._wait_for_handler()
+            await tab.close()
+            browser.stop()
+
+    @staticmethod
+    async def _click_on_player(tab) -> bool:
+        try:
+            player = await tab.select("#movie_player", 10)
+        except asyncio.TimeoutError:
+            logger.warning(f'update failed: unable to locate video player on the page')
+            return False
+        else:
+            await player.click()
+            return True
+
+    async def _wait_for_handler(self) -> bool:
+        try:
+            await asyncio.wait_for(self._extraction_done.wait(), timeout=30)
+        except asyncio.TimeoutError:
+            logger.warning(f'update failed: timeout waiting for outgoing API request')
+            return False
+        else:
+            logger.info('update was succeessful')
+            return True
+
+    async def _send_handler(self, event: nodriver.cdp.network.RequestWillBeSent):
```

Comment:
Return notation, not exactly required by some parts of the code has None notated and others don't, prob worth notating the same way consistently